File: v1/src/main/java/com/api/v1/controllers/customer/CustomerController.java
package com.api.v1.controllers.customer;

import com.api.v1.dtos.requests.CustomerRegistrationRequestDto;
import com.api.v1.dtos.responses.CustomerResponseDto;
import com.api.v1.services.customer.CustomerRegistrationService;
import com.api.v1.services.customer.CustomerUpdateService;
import com.api.v1.services.customer.CustomersDeletionService;
import com.api.v1.services.customer.CustomersRetrieveService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("api/v1/customers")
public class CustomerController {

    @Autowired
    private CustomerRegistrationService registrationService;

    @Autowired
    private CustomersDeletionService deletionService;

    @Autowired
    private CustomersRetrieveService retrieveService;

    @Autowired
    private CustomerUpdateService updateService;

    @PostMapping
    @ResponseStatus(value = HttpStatus.CREATED)
    public Mono<CustomerResponseDto> register(@Valid @RequestBody CustomerRegistrationRequestDto request) {
        return registrationService.register(request);
    }

    @DeleteMapping
    public Mono<Void> deleteAll() {
        return deletionService.deleteAll();
    }

    @DeleteMapping("{ssn}")
    public Mono<Void> deleteBySsn(@PathVariable String ssn) {
        return deletionService.deleteBySsn(ssn);
    }

    @GetMapping
    @ResponseStatus(value = HttpStatus.OK)
    public Flux<CustomerResponseDto> retrieveAll() {
        return retrieveService.retrieveAll();
    }

    @GetMapping("{ssn}")
    @ResponseStatus(value = HttpStatus.OK)
    public Mono<CustomerResponseDto> retrieveBySsn(@PathVariable String ssn) {
        return retrieveService.retrieveBySsn(ssn);
    }

    @PutMapping
    @ResponseStatus(value = HttpStatus.OK)
    public Mono<CustomerResponseDto> update(
            @Valid @RequestBody CustomerRegistrationRequestDto request
    ) {
        return updateService.update(request);
    }
    
}

File: kotlin/v2/src/main/kotlin/com/api/v2/customer/controller/CustomerController.kt
package com.api.v2.customer.controller

import com.api.v2.customer.anotations.SSN
import com.api.v2.customer.service.CustomerRegistrationService
import com.api.v2.customer.dtos.CustomerResponseDto
import com.api.v2.customer.dtos.CustomerModificationRequestDto
import com.api.v2.customer.dtos.CustomerRegistrationRequestDto
import com.api.v2.customer.service.CustomerDeletionService
import com.api.v2.customer.service.CustomerModificationService
import com.api.v2.customer.service.CustomerRetrievalService
import jakarta.validation.Valid
import kotlinx.coroutines.flow.Flow
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("api/v2/customers")
class CustomerController {

    @Autowired
    private lateinit var customerRegistrationService: CustomerRegistrationService

    @Autowired
    private lateinit var customerModificationService: CustomerModificationService

    @Autowired
    private lateinit var customerRetrievalService: CustomerRetrievalService

    @Autowired
    private lateinit var customerDeletionService: CustomerDeletionService

    @PostMapping
    @ResponseStatus(value = HttpStatus.CREATED)
    suspend fun register(@RequestBody requestDto: @Valid CustomerRegistrationRequestDto): CustomerResponseDto {
        return customerRegistrationService.register(requestDto)
    }

    @PutMapping("{ssn}")
    @ResponseStatus(value = HttpStatus.NO_CONTENT)
    suspend fun modify(
        @PathVariable ssn: @SSN String,
        @RequestBody requestDto: @Valid CustomerModificationRequestDto
    ): CustomerResponseDto {
        return customerModificationService.modify(ssn, requestDto)
    }

    @GetMapping("{ssn}")
    @ResponseStatus(value = HttpStatus.OK)
    suspend fun findBySsn(@PathVariable ssn: @SSN String): CustomerResponseDto {
        return customerRetrievalService.findBySsn(ssn)
    }

    @GetMapping
    @ResponseStatus(value = HttpStatus.OK)
    suspend fun findAll(): Flow<CustomerResponseDto> {
        return customerRetrievalService.findAll()
    }

    @DeleteMapping
    @ResponseStatus(value = HttpStatus.NO_CONTENT)
    suspend fun deleteAll() {
        return customerDeletionService.deleteAll()
    }

    @DeleteMapping("{ssn}")
    @ResponseStatus(value = HttpStatus.NO_CONTENT)
    suspend fun deleteBySsn(@PathVariable ssn: @SSN String) {
        return customerDeletionService.deleteBySsn(ssn)
    }

}